kernel/src/expressions/scalars.rs

```diff
@@ -49,8 +60,14 @@ impl Scalar {
             Self::Binary(_) => DataType::BINARY,
             Self::Decimal(_, precision, scale) => DataType::decimal_unchecked(*precision, *scale),
             Self::Null(data_type) => data_type.clone(),
+            Self::Struct(_, fields) => DataType::struct_type(fields.clone()),
```

Comment:
If I understand correctly, the `Vec<StructFields>` we store is redundant (and thus could get out of sync with the `Vec<Scalar>` we also store)? If we only had the scalars, I think we can do here:
```suggestion
            Self::Struct(scalars) => DataType::struct_type(
                scalars.iter().map(|s| s.data_type()).collect()),
```